constrain-lib/src/main/java/org/gradle/dependency/constrain/lib/serialize/ConstraintsJsonReader.java

```diff
@@ -0,0 +1,327 @@
+            try (BufferedReader useIn = new BufferedReader(new InputStreamReader(in))) {
+                useIn.mark(1 << 24); // Mark the stream so we can reset it later
+                // StreamReadFeature.AUTO_CLOSE_SOURCE was disabled when creating the ObjectFactory
+                json = mapper.readTree(useIn);
+                useIn.reset();
+                inputLines = useIn.lines().collect(Collectors.toList());
+                if (inputLines.isEmpty()) {
+                    throw new DependencyConstrainException("File is empty");
+                }
+            } catch (IOException ex) {
+                throw new DependencyConstrainException("Unable to read dependency constraints", ex);
+            }
+            return new InputStreamExtractedData(json, inputLines);
+        }
+    }
+
+    private static List<String> generateFormattedJson(ObjectMapper objectMapper, JsonNode json) {
+        final DefaultPrettyPrinter prettyPrinter = new DefaultPrettyPrinter();
+        monkeyPatchDefaultPrettyPrinter(prettyPrinter);
+        final DefaultIndenter indenter = new DefaultIndenter();
+        prettyPrinter.indentArraysWith(indenter);
+        prettyPrinter.indentObjectsWith(indenter);
+        final ObjectWriter writer = objectMapper.writer(prettyPrinter);
+        final String output;
+        try {
+            output = writer.writeValueAsString(json);
+        } catch (JsonProcessingException e) {
+            throw new DependencyConstrainException(
+                "Unable to generate formatted JSON for dependency constraints",
+                e
+            );
+        }
+        return Arrays.asList(output.split(System.lineSeparator()));
+    }
+
+    /**
+     * Verifies that the input matches exactly the formatting that we expect.
+     * Will generate an exception containing the diff required to make the input match the expected format.
+     */
+    private static void verifyNoFormattingDifferences(List<String> input, List<String> formatted) {
+        Patch<String> patch = DiffUtils.diff(input, formatted);
+        if (!patch.getDeltas().isEmpty()) {
+            final List<String> strings = UnifiedDiffUtils.generateUnifiedDiff(
+                DEPENDENCY_CONSTRAINTS_FILE_PATH_RELATIVE,
+                DEPENDENCY_CONSTRAINTS_FILE_PATH_RELATIVE,
+                input,
+                patch,
+                0
+            );
+            String newline = System.lineSeparator();
+            final String fixDiffToApply =
+                strings
+                    .stream()
+                    .collect(Collectors.joining(newline + "  ", "  ", ""));
+            throw new DependencyConstrainException(
+                "The dependency constraints are not formatted correctly. Please apply this patch to fix the formatting:" +
+                    newline +
+                    fixDiffToApply
+            );
+        }
+    }
+
+    private static void monkeyPatchDefaultPrettyPrinter(DefaultPrettyPrinter prettyPrinter) {
```

Comment:
Why is this necessary? It is a protected field, just subclass it and don't do any reflection shenanigans.